File: src/cgeohash_fn_repeaters.cpp
#include <v8.h>
#include <node.h>
#include "../includes/cvv8/detail/convert_core.hpp"

#include "cgeohash.hpp"
#include "cgeohash_fn_repeaters.hpp"
#include "cgeohash_nanoseconds.hpp"

#include <iostream>

#define _THROW_NODE_ERROR(MSG) ThrowException(v8::Exception::Error(v8::String::New(MSG)));


namespace cgeohash {

	// Node.JS Hooks to GeoHash encoding
	v8::Handle<v8::Value> encode_fn_repeater(const v8::Arguments& args) {
	    v8::HandleScope scope;

			for(int i = 0; i < 1e6;  i++) {
				encode(37.8324, 112.5584, 9);
			}
			
			return scope.Close(v8::Undefined());
	}

	v8::Handle<v8::Value> decode_fn_repeater(const v8::Arguments& args) {
	    v8::HandleScope scope;

	    if (args.Length() < 2) {
	        return _THROW_NODE_ERROR("Takes 1 parameters: num_times, hash_string");
	    }

		int i = 0;
		const int num_times = cvv8::CastFromJS< int >(args[i++]);
			const std::string hash_string = cvv8::CastFromJS<std::string>(args[i++]);
      if (hash_string.empty()) {
          return _THROW_NODE_ERROR("Parameter 0 must be a valid string");
      }

			const uint64_t _nanoseconds = nanoseconds();
			for(int i = 0; i < num_times;  i++) {
				// std::cout << __FILE__ << ':' << __LINE__ << '=' << i << std::endl;
		    decode(hash_string);
			}
			return scope.Close(cvv8::CastToJS<uint64_t>((nanoseconds() - _nanoseconds) / num_times));
			// const double   _seconds = seconds_differience_of_nanoseconds(_nanoseconds) / num_times;
			// return scope.Close(cvv8::CastToJS<double>(_seconds));
	}
	
	v8::Handle<v8::Value> decode_bbox_fn_repeater(const v8::Arguments& args) {
    v8::HandleScope scope;

    if (args.Length() < 2) {
        return _THROW_NODE_ERROR("Takes 1 parameters: num_times, hash_string");
    }

	int i = 0;
	const int num_times = cvv8::CastFromJS< int >(args[i++]);
		const std::string hash_string = cvv8::CastFromJS<std::string>(args[i++]);
    if (hash_string.empty()) {
        return _THROW_NODE_ERROR("Parameter 1 must be a valid string");
    }

		const uint64_t _nanoseconds = nanoseconds();
		for(int i = 0; i < num_times;  i++) {
			// std::cout << __FILE__ << ':' << __LINE__ << '=' << i << std::endl;
	    decode_bbox(hash_string);
		}
		return scope.Close(cvv8::CastToJS<uint64_t>((nanoseconds() - _nanoseconds) / num_times));
		// const double   _seconds = seconds_differience_of_nanoseconds(_nanoseconds) / num_times;
		// return scope.Close(cvv8::CastToJS<double>(_seconds));
	}

	v8::Handle<v8::Value> neighbor_fn_repeater(const v8::Arguments& args) {
	    v8::HandleScope scope;

	    if (args.Length() < 3) {
	        return _THROW_NODE_ERROR("Takes 3 parameters: num_times, hash_string, and direction []");
	    }

		int i = 0;
		const int num_times = cvv8::CastFromJS< int >(args[i++]);
			const std::string hash_string = cvv8::CastFromJS<std::string>(args[i++]);
      if (hash_string.empty()) {
          return _THROW_NODE_ERROR("Parameter 1 must be a valid string");
      }

			const std::list<uint32_t> directions = cvv8::CastFromJS< std::list<uint32_t> >(args[i++]);
			if (directions.size() != 2) {
          return _THROW_NODE_ERROR("Parameter 2 must be an array with 2 numbers");
      }
			
			const int directions_array [] = {
				directions.front(), // Only 2 elements
				directions.back()   // Only 2 elements
			};

			const uint64_t _nanoseconds = nanoseconds();
			for(int i = 0; i < num_times;  i++) {
				// std::cout << __FILE__ << ':' << __LINE__ << '=' << i << std::endl;
				neighbor(hash_string, directions_array);
			}
			return scope.Close(cvv8::CastToJS<uint64_t>((nanoseconds() - _nanoseconds) / num_times));
			// const double   _seconds = seconds_differience_of_nanoseconds(_nanoseconds) / num_times;
			// return scope.Close(cvv8::CastToJS<double>(_seconds));
	}

	void register_node_fn_repeaters(v8::Handle<v8::Object> target) {
    node::SetMethod(target, "encode_fn_repeater",      encode_fn_repeater);
    node::SetMethod(target, "decode_fn_repeater",      decode_fn_repeater);
    node::SetMethod(target, "decode_bbox_fn_repeater", decode_bbox_fn_repeater);
    node::SetMethod(target, "neighbor_fn_repeater",    neighbor_fn_repeater);
	}

}

File: lib/cgeohash_fn_repeaters.js
//
// Map the cgeohash C++ functions to callbacks to repeat X-Many times
// 
// Ex:
// var repeat_in_js  = require('cgeohash_fn_repeaters').repeat_in_js
// var repeat_in_cpp = require('cgeohash_fn_repeaters').repeat_in_cpp
// var avg_seconds_per_decode = {
//   looping_in_js:  repeat_in_js.decode( 1000*1000, 'abcd'),
//   looping_in_cpp: repeat_in_cpp.decode(1000*1000, 'abcd'),
// }
// 
// console.log('Seconds per call Node <-> C++ = ' + looping_in_js);
// console.log('Seconds per call C++  <-> C++ = ' + looping_in_cpp);
//

var cgeohash                 = require('../build/Release/cgeohash');
var cgeohash_fn              = require('./cgeohash_fn');
var repeats_callback_wrapper = require('v8-profiler-table').repeats_callback_wrapper;

// Alias the <...>_obj methods to the expected <...> method names
// Loop in JS and record the seconds / call
var cgeohash_fn_repeat_js = {
	encode: function(num_times, latitude, longitude, numberOfChars) {
		var loop = repeats_callback_wrapper(num_times, function() {
			return cgeohash_fn.encode(latitude, longitude, numberOfChars);
		});
		return loop();
	},

	decode: function(num_times, hash_string) {
		var loop = repeats_callback_wrapper(num_times, function() {
			return cgeohash_fn.decode(hash_string);
		});
		return loop();
	},

	decode_bbox: function(num_times, hash_string) {
		var loop = repeats_callback_wrapper(num_times, function() {
			return cgeohash_fn.decode_bbox(hash_string);
		});
		return loop();
	},

	neighbor: function(num_times, hash_string, direction) {
		var loop = repeats_callback_wrapper(num_times, function() {
			return cgeohash_fn.neighbor(hash_string, direction);
		});
		return loop();
	},
};

// Alias the <...>_obj methods to the expected <...> method names
// Loop in CPP and record the seconds / call
var cgeohash_fn_repeat_cpp = {
    encode:        cgeohash.encode_fn_repeater,
    decode:        cgeohash.decode_fn_repeater,
    decode_bbox:   cgeohash.decode_bbox_fn_repeater,
    neighbor:      cgeohash.neighbor_fn_repeater,
};


module.exports = {
	repeat_in_js:  cgeohash_fn_repeat_js,
	repeat_in_cpp: cgeohash_fn_repeat_cpp,
};
